core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -215,4 +434,109 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        if (isTermsLookupQuery()) {
+            terms = context.handleTermsLookup(termsLookup.toTermsLookup(context));
+        } else {
+            terms = new ArrayList<>();
+            terms.addAll(values);
+        }
+        if (terms.isEmpty()) {
+            return Queries.newMatchNoDocsQuery();
+        }
+        return handleTermsQuery(terms, fieldName, context, minimumShouldMatch, disableCoord);
+    }
+
+    static Query handleTermsQuery(List<Object> terms, String fieldName, QueryShardContext context, String minimumShouldMatch, boolean disableCoord) {
+        MappedFieldType fieldType = context.fieldMapper(fieldName);
+        String indexFieldName;
+        if (fieldType != null) {
+            indexFieldName = fieldType.names().indexName();
+        } else {
+            indexFieldName = fieldName;
+        }
+
+        Query query;
+        if (context.isFilter()) {
+            if (fieldType != null) {
+                query = fieldType.termsQuery(terms, context);
+            } else {
+                BytesRef[] filterValues = new BytesRef[terms.size()];
+                for (int i = 0; i < filterValues.length; i++) {
+                    filterValues[i] = BytesRefs.toBytesRef(terms.get(i));
+                }
+                query = new TermsQuery(indexFieldName, filterValues);
+            }
+        } else {
+            BooleanQuery bq = new BooleanQuery(disableCoord);
+            for (Object term : terms) {
+                if (fieldType != null) {
+                    bq.add(fieldType.termQuery(term, context), BooleanClause.Occur.SHOULD);
+                } else {
+                    bq.add(new TermQuery(new Term(indexFieldName, BytesRefs.toBytesRef(term))), BooleanClause.Occur.SHOULD);
+                }
+            }
+            Queries.applyMinimumShouldMatch(bq, minimumShouldMatch);
+            query = bq;
+        }
+        return query;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (this.fieldName == null) {
+            validationException = addValidationError("field name cannot be null.", validationException);
+        }
+        if (isTermsLookupQuery()) {
+            QueryValidationException exception = termsLookup.validate();
+            if (exception != null) {
+                validationException = QueryValidationException.addValidationErrors(exception.validationErrors(), validationException);
+            }
+        }
+        return validationException;
+    }
+
+    @Override
+    protected TermsQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        TermsQueryBuilder termsQueryBuilder = new TermsQueryBuilder(in.readString());
+        if (in.readBoolean()) {
+            termsQueryBuilder.termsLookup = TermsLookupBuilder.readTermsLookupFrom(in);
+        } else {
+            termsQueryBuilder.values.addAll((List<Object>) in.readGenericValue());
+        }
```

Comment:
I'm conflicted here, not sure if this either/or serialization is problematic on the long run. I understand that having both values and termsLookup makes no sense, we should also catch it before serialization etc... but I think it would be safer to potentially serialize both field, e.g. if for some reason we construct an "invalid" q.b. with both options and don't catch it, serializing it will favour the termsLookup and drop the other, so the query is valid afterwards. That might be paranoid on my side but I think its safer at the cost of some extra bits.